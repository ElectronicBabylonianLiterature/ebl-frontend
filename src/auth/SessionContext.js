import React from 'react'
import Session from './Session'

const SessionContext = React.createContext(new Session('', '', 0, []))

export default SessionContext
