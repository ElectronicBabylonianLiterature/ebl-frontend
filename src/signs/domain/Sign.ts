export default interface Sign {
  name: string
  lists: readonly SignListRecord[]
  values: readonly Value[] | null
  logograms: readonly Logogram[]
  mesZl: string | null
}
interface Logogram {
  logogram: string
  atf: string
  wordId: readonly string[]
  schrammLogogramme: string
}
interface Value {
  value: string
  subIndex: number | null
}

interface SignListRecord {
  name: string
  number: string
}

export interface SignQuery {
  value: string
  subIndex: string
  isIncludeHomophones: boolean
  isComposite: boolean
  signList: string
}
