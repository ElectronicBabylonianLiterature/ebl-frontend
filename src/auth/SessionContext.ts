import React, { Context } from 'react'
import { guestSession, Session } from 'auth/Session'

const SessionContext: Context<Session> = React.createContext<Session>(
  guestSession
)

export default SessionContext
