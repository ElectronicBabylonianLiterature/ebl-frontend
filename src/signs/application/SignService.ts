import Bluebird from 'bluebird'
import Sign, { SignQuery } from 'signs/domain/Sign'
import SignRepository from 'signs/infrastructure/SignRepository'
import { AnnotationToken } from 'fragmentarium/domain/annotation-token'

import { CroppedAnnotation } from 'signs/domain/CroppedAnnotation'

export default class SignService {
  private readonly signsRepository: SignRepository

  constructor(signsRepository: SignRepository) {
    this.signsRepository = signsRepository
  }

  getImages(signName: string): Bluebird<CroppedAnnotation[]> {
    return this.signsRepository.getImages(signName)
  }

  associateSigns(
    tokens: ReadonlyArray<ReadonlyArray<AnnotationToken>>
  ): Bluebird<ReadonlyArray<ReadonlyArray<AnnotationToken>>> {
    return this.signsRepository.associateSigns(tokens)
  }

  search(signQuery: SignQuery): Bluebird<Sign[]> {
    return this.signsRepository.search(signQuery)
  }

  find(signName: string): Bluebird<Sign> {
    return this.signsRepository.find(signName)
  }
}
