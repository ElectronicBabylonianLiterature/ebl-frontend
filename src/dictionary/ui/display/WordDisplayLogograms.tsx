import React, { Fragment } from 'react'
import Bluebird from 'bluebird'
import { Col, Row } from 'react-bootstrap'
import SignService from 'signs/application/SignService'
import { Markdown } from 'common/Markdown'
import withData from 'http/withData'
import Sign from 'signs/domain/Sign'
import { flatten } from 'lodash'

function filterLogogramsData(
  signs: readonly Sign[],
  wordId: string
): { firstSignName: string; unicode: string; note: string }[] {
  return flatten(
    signs.map((sign) =>
      sign.logograms
        .filter((logogram) => logogram.wordId.includes(wordId))
        .map((logogram) => {
          return {
            firstSignName: sign.name,
            unicode: sign.displayCuneiformSigns,
            note: logogram.schrammLogogramme,
          }
        })
    )
  )
}

function LogogramsDisplay({
  signs,
  wordId,
}: {
  signs: readonly Sign[]
  wordId: string
}): JSX.Element {
  const filteredLogograms = filterLogogramsData(signs, wordId)
  return (
    <Fragment key="akkadischeLogogramme">
      {filteredLogograms.map((logogram, i) => (
        <Row className="ml-5" key={`logogram_${i}`}>
          <Col>
            <a href={`/signs/${logogram.firstSignName}`}>{logogram.unicode}</a>
            &emsp;
            <Markdown text={logogram.note} />
          </Col>
        </Row>
      ))}
    </Fragment>
  )
}

export default withData<
  {
    signService: SignService
    wordId: string
  },
  { wordId: string },
  Sign[]
>(
  ({ data: signs, wordId, ...props }) => {
    console.log(signs)
    return <LogogramsDisplay signs={signs} wordId={wordId} {...props} />
  },
  ({ signService, wordId }) => {
    return Bluebird.all(
      signService.search({ wordId: decodeURIComponent(wordId) })
    )
  }
)
