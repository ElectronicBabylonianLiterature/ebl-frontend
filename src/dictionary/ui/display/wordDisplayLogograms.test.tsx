import React, { ReactNode } from 'react'
import { render, screen } from '@testing-library/react'
import { MemoryRouter, Route, RouteComponentProps } from 'react-router-dom'
import SessionContext from 'auth/SessionContext'
import MemorySession from 'auth/Session'

import WordDisplayLogograms from 'dictionary/ui/display/WordDisplayLogograms'
import SignService from 'signs/application/SignService'
import Bluebird from 'bluebird'
import { sign } from 'signs/domain/Sign.test'

jest.mock('signs/application/SignService')
const signService = new (SignService as jest.Mock<jest.Mocked<SignService>>)()

const session = new MemorySession(['read:words'])

let container: HTMLElement

describe('Fetch logograms', () => {
  beforeEach(async () => {
    signService.search.mockReturnValue(Bluebird.resolve([sign]))
    renderWordDisplayLogograms()
    await screen.findByText('some notes')
    expect(signService.search).toBeCalledWith({
      wordId: sign.logograms[0].wordId[0],
    })
  })
  it('correctly displays unicode', async () => {
    await screen.findAllByText(new RegExp(sign.logograms[0].unicode))
    expect(container).toMatchSnapshot()
  })
})

function renderWordDisplayLogograms() {
  container = render(
    <MemoryRouter initialEntries={['/dictionary/id']}>
      <SessionContext.Provider value={session}>
        <Route
          path="/dictionary/:id"
          render={(props: RouteComponentProps<{ id: string }>): ReactNode => (
            <WordDisplayLogograms
              wordId={sign.logograms[0].wordId[0]}
              signService={signService}
              {...props}
            />
          )}
        />
      </SessionContext.Provider>
    </MemoryRouter>
  ).container
}
