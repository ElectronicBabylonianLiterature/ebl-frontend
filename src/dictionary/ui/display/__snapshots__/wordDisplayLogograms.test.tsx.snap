// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch logograms correctly displays unicode 1`] = `
<div>
  <div
    class="ml-5 row"
  >
    <div
      class="col"
    >
      <span
        class="unicode-large"
      >
        <a
          href="/signs/BU"
        >
          íÅç
        </a>
      </span>
      ‚ÄÉ
      <span
        class=""
      >
        <em>
          GID‚ÇÇ
        </em>
         some notes
      </span>
    </div>
  </div>
  <div
    class="text-center border border-dark mx-5 m-3 p-2"
  >
    <strong>
      From
    </strong>
    <br />
    Schramm, W.
    , 
    <em>
      Akkadische Logogramme
    </em>
    . 
    Zweite, revidierte Auflage. G√∂ttinger Beitr√§ge zum Alten Orient 5. G√∂ttingen: Universit√§tsverlag
G√∂ttingen, ¬≤2010
    .
    <br />
    <a
      class="text-dark "
      href="https://creativecommons.org/licenses/by-nd/3.0/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <strong>
        CC BY-ND 3.0
      </strong>
    </a>
    <br />
    <a
      class="text-dark "
      href="https://univerlag.uni-goettingen.de/handle/3/isbn-978-3-941875-65-4"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        class="pointer__hover my-2 fas fa-external-link-square-alt"
      />
    </a>
  </div>
</div>
`;
